network/p2p/unicast/rate_limited_peers_map.go

```diff
@@ -0,0 +1,100 @@
+package unicast
+
+import (
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/peer"
+)
+
+type rateLimitedPeerMapItem struct {
+	lastActive time.Time
+}
+
+// rateLimitedPeersMap concurrency safe struct that stores rate limited peers in a map.
+// Map keys will expire or be deleted from the map based on the configured TTL.
+type rateLimitedPeersMap struct {
+	mu              sync.Mutex
+	ttl             time.Duration
+	cleanupInterval time.Duration
+	peers           map[peer.ID]*rateLimitedPeerMapItem
+	done            chan struct{}
+}
+
+func newRateLimitedPeersMap(ttl, cleanupInterval time.Duration) *rateLimitedPeersMap {
+	return &rateLimitedPeersMap{
+		mu:              sync.Mutex{},
+		ttl:             ttl,
+		cleanupInterval: cleanupInterval,
+		peers:           make(map[peer.ID]*rateLimitedPeerMapItem),
+		done:            make(chan struct{}),
+	}
+}
+
+// get returns ok if peerID key exists in map. This func will also update the lastActive
+// timestamp for the item refreshing its ttl.
+func (r *rateLimitedPeersMap) exists(peerID peer.ID) bool {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if item, ok := r.peers[peerID]; ok {
+		item.lastActive = time.Now()
```

Comment:
nit: not sure if exists should update it.  It look like if peer gets limited, then any non-zero rate will keep it blocked forever.

Reply:
This will not happen because we are periodically cleaning up the expired keys.

Reply:
re: Everything is synchronized with the mutex so this would never happen. 

> this problem still remains: if peer gets ratelimited `IsRateLimited` will short-circuit the `MessageAllowed`/`BandwidthAllowed` checks.

Reply:
I think there are a few of concepts here that I think are causing confusion.

1. The rate limit package articulates limits in `units` per second
2. The `ttl` used for `rateLimitedPeersMap` defaults of 10 minutes. This implies that if a peer exceeds their limit, they are "locked out" for 10 minutes. As mentioned in my previous comment, if that peer attempts to connect again with blocked, their lockout time resets, so they're blocked for another 10 minutes.
3. There are 2 ways rate limits are used in this PR. I'll distinguish them as applying (adding the `units` to the rate limit and seeing if that exceeded the limit) vs enforcing (blocking requests from a peer that exceeded their limit)
    * When a message is received, we check if the limiter will allow the additional message/bandwidth. Here we have all of the information we need to apply and enforce the limit.
    * When a peer connects, we check if they are already rate limited. Here we don't have enough information to apply the limits (# message/bytes), but we have enough to enforce an existing limit. We just want an efficient way to determine if the peer should be allowed to connect at all.

A couple thoughts on this:
* I don't think we should be updating the timestamp when enforcing the limit since we don't actually have enough information yet, and it can create a permanent lockout. Instead, we should just check if the peer is locked out in `IsRateLimited()`, and update the timestamps in `Allowed()`. That way a peer can become locked out for a single `ttl`, but they will eventually be allowed again.
* I'm generally OK with the idea of locking a peer out for some period of time after exceeding the limit. I think it should be configurable in case we find that too high/low of a limit is problematic on mainnet.

Reply:
The TTL has nothing to do with the actual rate limits. It's only used to remove stale keys from the map so that unnecessary memory is not used.

Say rate limit is 5 messages per second and you send 6 messages and get rate limited. All messages after the 5th message will be **_blocked for 1 second_** . After 1 second has passed the next message that is sent will be allowed and the peer will be removed from the rate limited peers map see [here](https://github.com/onflow/flow-go/blob/khalil/1716-unicast-rate-limiter/network/p2p/unicast/message_rate_limiter.go#L50) and [here](https://github.com/onflow/flow-go/blob/khalil/1716-unicast-rate-limiter/network/p2p/unicast/bandwidth_rate_limiter.go#L50). TTL is only relevant to the cleanup loop. If lets say a user is rate limited and then becomes inactive (doesn't send any messages) that peer will be automatically "un rate limited" after the configured TTL has passed.